core/src/main/java/org/apache/iceberg/avro/AssignIds.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.avro;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import java.util.List;
+import java.util.Map;
+import org.apache.avro.Schema;
+
+/**
+ * Assign field ids to the Avro schema using the given field name to field id map
+ */
+public class AssignIds extends AvroSchemaVisitor<Schema> {
+  // Mapping constructed from table properties
+  private final Map<String, Integer> nameToId;
+  // To be used for maps and list. We do not store ids for maps and lists in our mapping, and it is not
+  // really required. For maps and list we assign a default id.
+  private static final int DEFAULT_ID = -1;
```

Comment:
I think mappings need to include IDs for map keys, values and list elements. Otherwise, those can't be projected unless they contain structs with projected fields.